### DISCUSSION TITLE
Upgrading Issue from 5.1.4 to 5.4.0 in PHP 7.4: Problems with PUT Data Retrieval in application/x-www-form-urlencoded ContentType

### DISCUSSION THREAD

**Discussion by baohanddd** [phalcon/cphalcon#16513] [General]
I encountered a problem after upgrading from 5.1.4 to 5.4.0 on PHP 7.4. I am unable to retrieve PUT data, which is always empty, but reverting back to version 5.1.4 resolves the issue. Currently, the project is using the ContentType application/x-www-form-urlencoded.

> var_dump($_SERVER['HTTP_CONTENT_TYPE']);
var_dump($this->request->getPut()); 
exit;

The Result:
> string(33) "application/x-www-form-urlencoded"
> array(0) {
> }

I looked into the code for both 5.1.4 and 5.4.0 and found the issue seems to be with the getPatchPut() function.

**getPatchPut in 5.4.0:**

![getPatchPut in 5.4.0](https://github.com/phalcon/cphalcon/assets/3166072/3ddb1245-af43-4305-a56c-91f3ca87b30c)

I'm not sure if version 5.4.0 no longer supports application/x-www-form-urlencoded, or is there a way to make PUT work under x-www-form-urlencoded?

**Comment by niden** [comment]
This is odd. I checked the code and that method was introduced with 5.1 and has not changed since then.

Can you try setting the `CONTENT_TYPE` instead of the `HTTP_CONTENT_TYPE` and see if that works?

**Reply by baohanddd** [reply]
The issue might be here. 

I have checked the differences in code versions on GitHub. Prior to version 5.2.3, the code behaved such that even if the content-type was 'application/x-www-form-urlencoded', it would enter the 'else' branch and execute the 'getRawBody()' function. 
![v5 2 3](https://github.com/phalcon/cphalcon/assets/3166072/8728a29e-b964-4d5e-97e3-8b265e38bba5)

The condition in the code from version 5.3.0 and later has been modified.
![v5 3 0](https://github.com/phalcon/cphalcon/assets/3166072/a4ed4d55-00dc-40e6-99dd-2e631a4283a0)

As long as the content-type is a string, and it is not 'multipart/form-data' or 'json', there will be no chance to retrieve data from the 'getRawBody()' function anymore.

**Reply by niden** [reply]
This is the related issue for that change

https://github.com/phalcon/cphalcon/issues/16271

I will check both out when I get back from Church.

**Reply by niden** [reply]
Thank you @baohanddd

**Reply by baohanddd** [reply]
Great Job, thanks 👍🏻

**Comment by niden** [comment]
@baohanddd This has been resolved in https://github.com/phalcon/cphalcon/pull/16520

Thanks to @sinbadxiii
